File: packages/app/android/src/flutter/java/io/invertase/firebase/common/FlutterFirebasePlugin.java
package io.invertase.firebase.common;

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */


import android.app.Activity;
import android.content.Context;
import android.os.Bundle;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry.Registrar;

public class FlutterFirebasePlugin {
  public Registrar registrar;

  public static String getChannelName(String module) {
    return "io.invertase.firebase/" + module;
  }

  public FlutterFirebasePlugin(Registrar registrar) {
    this.registrar = registrar;
  }

  public Context getContext() {
    return registrar.context();
  }

  public Activity getActivity() {
    return registrar.activity();
  }

  public Context getApplicationContext() {
    return getContext().getApplicationContext();
  }

  public void rejectResultWithExceptionMap(MethodChannel.Result result, Exception exception) {
    // TODO handle null?
    Map<String, String> exceptionMap = new HashMap<>(4);
    String code = "unknown";
    String message = exception.getMessage();
    exceptionMap.put("code", code);
    exceptionMap.put("nativeErrorCode", code); // TODO Needed?
    exceptionMap.put("message", message);
    exceptionMap.put("nativeErrorMessage", message); // TODO Needed?
    result.error(code, message, exceptionMap);
  }

  public static Bundle createBundleFromMap(Map<String, Object> map) {
    if (map == null) {
      return null;
    }

    Bundle bundle = new Bundle();
    for (Map.Entry<String, Object> jsonParam : map.entrySet()) {
      final Object value = jsonParam.getValue();
      final String key = jsonParam.getKey();
      if (value instanceof String) {
        bundle.putString(key, (String) value);
      } else if (value instanceof Integer) {
        bundle.putInt(key, (Integer) value);
      } else if (value instanceof Long) {
        bundle.putLong(key, (Long) value);
      } else if (value instanceof Double) {
        bundle.putDouble(key, (Double) value);
      } else if (value instanceof Boolean) {
        bundle.putBoolean(key, (Boolean) value);
      } else {
        throw new IllegalArgumentException(
          "Unsupported value type: " + value.getClass().getCanonicalName());
      }
    }
    return bundle;
  }

}

File: packages/app/android/src/flutter/java/io/invertase/firebase/common/FlutterFirebasePlugin.kt
package io.invertase.firebase.common

/*
 * Copyright (c) 2016-present Invertase Limited & Contributors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this library except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */


import android.app.Activity
import android.content.Context
import android.os.Bundle
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry.Registrar
import java.util.*

open class FlutterFirebasePlugin(var registrar: Registrar) {

  val context: Context
    get() = registrar.context()

  val activity: Activity
    get() = registrar.activity()

  val applicationContext: Context
    get() = context.applicationContext

  fun rejectResultWithExceptionMap(result: MethodChannel.Result, exception: Exception?) {
    // TODO handle null?
    val exceptionMap = HashMap<String, String>(4)
    val code = "unknown"
    val message = exception?.message ?: "An unknown error occurred"
    exceptionMap["code"] = code
    exceptionMap["message"] = message
    result.error(code, message, exceptionMap)
  }

  companion object {

    fun getChannelName(module: String): String {
      return "io.invertase.firebase/$module"
    }

    fun createBundleFromMap(map: Map<String, Any?>?): Bundle? {
      if (map == null) {
        return null
      }

      val bundle = Bundle()
      for ((key, value) in map) {
        if (value == null) {
          bundle.putString(key, null)
        } else {
          when (value) {
            is String -> bundle.putString(key, value)
            is Int -> bundle.putInt(key, value)
            is Long -> bundle.putLong(key, value)
            is Double -> bundle.putDouble(key, value)
            is Boolean -> bundle.putBoolean(key, value)
            else -> throw IllegalArgumentException(
              "Unsupported value type: " + value.javaClass.canonicalName!!)
          }
        }
      }
      return bundle
    }
  }

}
